wolfcrypt/src/asn.c

```diff
@@ -18490,6 +18490,289 @@ int wc_ParseCertPIV(wc_CertPIV* piv, const byte* buf, word32 totalSz)
+    int ret = 0;
+    size_t pos = 0;
+    size_t lineLen = 0;
+    char* curLine = NULL;
+    char* ptr = NULL;
+
+    if (in == NULL || inLen <= 0 || in[inLen] != '\0') {
+        return ASN_PARSE_E;
+    }
+    nextHdr = (MimeHdr*)XMALLOC(sizeof(MimeHdr), NULL, DYNAMIC_TYPE_PKCS7);
+    nextParam = (MimeParam*)XMALLOC(sizeof(MimeParam), NULL,
+                                    DYNAMIC_TYPE_PKCS7);
+    if (nextHdr == NULL || nextParam == NULL) {
+        XFREE(nextHdr, NULL, DYNAMIC_TYPE_PKCS7);
+        XFREE(nextParam, NULL, DYNAMIC_TYPE_PKCS7);
+        return MEMORY_E;
+    }
+    XMEMSET(nextHdr, 0, (word32)sizeof(MimeHdr));
+    XMEMSET(nextParam, 0, (word32)sizeof(MimeParam));
+
+    curLine = XSTRTOK(in, "\r\n", &ptr);
+    if (curLine == NULL) {
+        XFREE(nextHdr, NULL, DYNAMIC_TYPE_PKCS7);
+        XFREE(nextParam, NULL, DYNAMIC_TYPE_PKCS7);
+        return ASN_PARSE_E;
+    }
+
+    while (curLine != NULL) {
+        /* Leftover from previous line, add params to previous header. */
+        if (curLine[0] == ' ' && curHdr) {
+            mimeType = MIME_PARAM;
+        }
+        else {
+            mimeType = MIME_HDR;
+        }
+        start = end = 0;
+        lineLen = XSTRLEN(curLine);
+
+        for (pos = 0; pos < lineLen; pos++) {
+            char cur = curLine[pos];
+
+            if (mimeStatus == MIME_NAMEATTR && ((cur == ':' &&
+                mimeType == MIME_HDR) || (cur == '=' &&
+                mimeType == MIME_PARAM))) {
+                mimeStatus = MIME_BODYVAL;
+                end = pos-1;
+                ret = wc_MIME_header_strip(curLine, &nameAttr, start, end);
+                if (ret) {
+                    XFREE(nameAttr, NULL, DYNAMIC_TYPE_PKCS7);
+                    XFREE(nextHdr, NULL, DYNAMIC_TYPE_PKCS7);
+                    XFREE(nextParam, NULL, DYNAMIC_TYPE_PKCS7);
+                    return ret;
+                }
+                start = pos+1;
+            }
+            else if (mimeStatus == MIME_BODYVAL && cur == ';') {
+                end = pos-1;
+                ret = wc_MIME_header_strip(curLine, &bodyVal, start, end);
+                if (ret) {
+                    XFREE(nameAttr, NULL, DYNAMIC_TYPE_PKCS7);
+                    XFREE(bodyVal, NULL, DYNAMIC_TYPE_PKCS7);
+                    XFREE(nextHdr, NULL, DYNAMIC_TYPE_PKCS7);
+                    XFREE(nextParam, NULL, DYNAMIC_TYPE_PKCS7);
+                    return ret;
+                }
+                if (mimeType == MIME_HDR) {
+                    nextHdr->name = nameAttr;
+                    nextHdr->body = bodyVal;
+                    nextHdr->next = curHdr;
+                    curHdr = nextHdr;
+                    nextHdr = (MimeHdr*)XMALLOC(sizeof(MimeHdr), NULL,
+                                                DYNAMIC_TYPE_PKCS7);
+                    if (nextHdr == NULL) {
+                        XFREE(nextParam, NULL, DYNAMIC_TYPE_PKCS7);
+                        return MEMORY_E;
```

Comment:
Looks like this error case could leak `nextHdr` and `nextParam`. I see a few other places as well. If you want to utilize a goto and have a single cleanup that would be okay.

Reply:
Good idea, that also cleans up the code.  Fixed.